### PR TITLE
MST-675: QuerySet Performance for  StudentOnboardingStatusByCourse Endpoint

### PR DESCRIPTION
**Pull request by MichaelRoytman**
**Description:**

See the corresponding edx-platform PR [here](https://github.com/edx/edx-platform/pull/26789).

The original use of `user.has_perm('can_take_proctored_exam')` had very poor performance. The `can_take_proctored_exam` permission, defined in the LMS, was not designed for use in bulk operations. It was being called for each user in a loop, resulting in many queries to the database. This change uses a new service endpoint that performs this permission checking in the database, resulting in one single query to the necessary LMS SQL tables and many fewer queries to the modulestore.

**JIRA:**

[MST-675](https://openedx.atlassian.net/browse/MST-675)

**Pre-Merge Checklist:**

- [ ] Updated the version number in `edx_proctoring/__init__.py` and `package.json` if these changes are to be released.
- [ ] Described your changes in `CHANGELOG.rst`
- [ ] Confirmed Github reports all automated tests/checks are passing.
- [ ] Approved by at least one additional reviewer.

**Post-Merge:**

- [ ] Create a tag matching the new version number.

### CODE REVIEW

File: edx_proctoring/api.py [hunk 1]
```diff
@@ -2490,12 +2490,13 @@ def get_integration_specific_email(provider):
     return getattr(provider, 'integration_specific_email', None) or constants.DEFAULT_CONTACT_EMAIL
 
 
-def get_enrollments_for_course(course_id):
+def get_enrollments_for_course_can_take_proctored_exams(course_id, text_search=None):
```
Comment (schenedx) [inline edx_proctoring/api.py:2493]: Why do we need the API function? Why not just use the `enrollment_service` directly in the view? I see patterns either way in edx_proctoring, so there is not standard on this, right?

**Review by schenedx — APPROVED** [review]
I think this looks much better.

File: edx_proctoring/api.py (hunk 1 above)
Comment (MichaelRoytman) [reply edx_proctoring/api.py]: I assume it's to provide an abstraction over how data is being fetched, allowing us to keep that logic out of the view.

File: edx_proctoring/tests/test_views.py [hunk 2]
```diff
@@ -855,182 +855,6 @@ def test_pagination_maintains_query_params(self):
         self.assertIn(text_search_string, next_url)
         self.assertIn(statuses_filter_string, next_url)
 
-    def test_partial_text_search_username(self):
```
Comment (schenedx) [inline edx_proctoring/tests/test_views.py:858]: Why were all these test cases removed?

Comment (MichaelRoytman) [reply edx_proctoring/tests/test_views.py:858]: edx_proctoring is no longer doing the username and email searching - it has been moved to the LMS as part of the `get_enrollments_can_take_proctored_exams ` method. I wrote tests in the LMS to capture this functionality.

Comment (schenedx) [reply edx_proctoring/tests/test_views.py:858]: Got it. But I feel like these tests also cover the restful API interface on the endpoint here. Therefore, keeping them here won't hurt. We should have double coverage here.

Comment (MichaelRoytman) [reply edx_proctoring/tests/test_views.py:858]: We're mocking out the enrollments service, so I would need to eitherreplicate the user searching there, which I don't think we want to do, or I'd need to import the function from the LMS here. Are you suggesting the second?

Comment (schenedx) [reply edx_proctoring/tests/test_views.py:858]: I see. So, if we keep mocking here, the tests are no longer relevant. 
That said, I believe you can update one of these tests to check the mock is called with text search passed into the mock. That way, you leave maybe 1 test to ensure the text search interface is implemented. I am not suggesting you need to import from LMS

**Review by schenedx — APPROVED** [review]
👍

**Pull request merged**